refactor(api): simplify logging

ojeytonwilliams opened the pull request:
This PR lets us easily add all the metadata that is currently logged in /status/ping. Any log call that uses `{ req }` will log the data we want. i.e. `req.log.debug({req}, 'message')` or `req.log.child({req}).debug('message')`.

Guidelines for usage:

request/response logging

```ts
// inside request handler
const logger = req.log.child({req});
// other code
logger.info('message')
```

logging outside of request handlers

```ts
fastify.log.debug('message')
```


This has changed a lot from my initial PR, thanks to Mrugesh's feedback, so:

## original PR description

This is closer to how fastify does it by default. The key differences are that the logs are always DEBUG level and we don't automatically log responses.

@raisedadead I removed the status specific stuff for a few reasons. Firstly, it's handled by the `onRequest` hook, but, more importantly, I think we should have a single source of truth for which parts of the request we log. As of this PR that's `requestSerializer`.

It seems like a good start, but if you think we need more, let me know and I can add it in. One thing we shouldn't do, though, is log all the cookies. They include `jwt_access_token` and we don't want that leaking out.

Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/how-to-open-a-pull-request/).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or Gitpod.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Closes #XXXXX

<!-- Feel free to add any additional description of changes below this line -->

ojeytonwilliams committed:
refactor: simplify logging

This is closer to how fastify does it by default. The key differences
are that the logs are always DEBUG level and we don't automatically
log responses

ShaunSHamilton reviewed (commented):
I seem to get duplicate logs of every request.

Is this expected, because strict mode React? Or, do we have two things logging the same thing now?

ojeytonwilliams commented:
React? And, no, you shouldn't get any duplicates. When I tried it, I got one log for the request and another if the route doesn't exist. You should only get more if you call a logger.

I'll double check in a bit, though.

ShaunSHamilton commented:
> React? And, no, you shouldn't get any duplicates.

I used the client to change the profile settings.

ojeytonwilliams commented:
Ah, I see. It's possible the client is messing up, but, regardless of how many requests it sends, there should be one request for each.

raisedadead commented on api/src/server.ts:
```diff
@@ -53,6 +53,10 @@ const start = async () => {
     genReqId: () => randomBytes(8).toString('hex'),
     disableRequestLogging: true
   });
+  fastify.addHook('onRequest', (req, _reply, done) => {
```
This reduces the granular control on 'specific' requests that we may want to log - example: I always want the ping endpoint, but never the others unless the 'debug' filter is on.

I see your point about single source of truth -- but this makes it "too noisy" even when we turn on "debug" filters.

I would like more control.

ojeytonwilliams replied:
Okay, what would you like to see at a given log-level?

What I was thinking was that we'd want a certain baseline when we set the log-level to debug. i.e. what `requestSerializer` gives us: `method`, `url`, `ip`, `hostname`, `remoteAddress` and `remotePort`.  Then, if you want more for a specific route, you'd just do this:

```
\\ in route handler:

req.log.debug({ more: "stuff" }, "a useful message").
```

and see:

```
[16:26:28 UTC] DEBUG: incoming request {"reqId":"d8773cb47f268617","req":{"method":"GET","url":"/status/ping","ip":"127.0.0.1","hostname":"localhost:3000","remoteAddress":"127.0.0.1","remotePort":36760}}
[16:26:28 UTC] DEBUG: a useful message {"reqId":"d8773cb47f268617","more":"stuff"}
```

raisedadead replied:
The thing is this hook as is will log every request for every end-point. That is noisy. Sure we could update the baseline in the serializer, but we do not need to add this hook. 

I would prefer adding logs to some/few of our endpoints to begin with.

ojeytonwilliams replied:
I think we'll want to work out how to filter logs by endpoint and then we can turn them all on.

I do see your point, though, and I'll get rid of the hook.

Is there extra stuff you'd like to see in the serializer? If so, how about I update it and use it for this endpoint?

raisedadead replied:
> Is there extra stuff you'd like to see in the serializer? If so, how about I update it and use it for this endpoint?

Sure - go ahead and add anything you see fit. Consolidation is better. I do not have any specific needs other than the end user IP

raisedadead commented on api/src/routes/public/status.ts:
```diff
@@ -15,25 +14,7 @@ export const statusRoute: FastifyPluginCallbackTypebox = (
-    const reqId = req.id || 'REQ_ID not found';
-    const headers = isEmpty(req.headers) ? 'HEADERS not found' : req.headers;
-    const ip =
-      req.headers['x-forwarded-for'] ||
-      req.headers['x-real-ip'] ||
-      req.ip ||
```
I do want to debug some of these information from time to time. So a child logger (which adds the request specific annotations automatically).

ojeytonwilliams committed:
Revert "refactor: simplify logging"

This reverts commit cdd3cf50f2e03ec467473cf517dbd2739a0acd3f.

ojeytonwilliams committed:
feat: update requestSerializer and use to log

ojeytonwilliams commented:
Round two @ShaunSHamilton @raisedadead.

I wasn't 100% sure what you wanted with respect to ip, x-real-ip and x-forwarded-for logs, Mrugesh, so this is my best effort.

I decided not to use `fastify.child`, not because it's wrong (it's not), but to show the alternative and because using `debug` directly is more concise.

`headers.cookies` contains `jwt_access_token`, so I redacted it entirely. We can remove specific cookies, if we want, but it's not as simple as just specifying `redact: "req.HEADERS.cookie.jwt_access_token"`, since redact only remove specific properties of objects. As such I haven't implemented it yet. Let me know if it's important.

ojeytonwilliams committed:
chore: remove comment

ojeytonwilliams committed:
fix: attach reqId to all req/reply logs

ojeytonwilliams committed:
refactor: simplify reqId generation

I was essentially re-creating the existing childLoggerFactory and
gaining precisely nothing.

ojeytonwilliams commented:
I know you're busy right now (and this is bikeshedding), so TL;DR: both approaches are equivalent for a single log request. `req.log.child({ req }).debug('message one')` gives the same output as `req.log.debug({ req }, 'message one')`. The difference is that you can reuse what `req.log.child({ req })` returns (i.e. what Shaun did in his PR).

So, I don't care which we use. If you like `.child`, I'm fine with it.

If you're interested in the bikeshedding, and for future reference, here's what I think is going on:

> The thing is the child logger automatically adds all the metadata from the request

Yes and no. "No" because `fastify.log.child({ req }).debug('message')` doesn't add `reqId`, because it doesn't have access to the it. "Yes" because behind the scenes it's a child logger that includes the `reqId`. We use that child logger whenever we log.

>  that is the reason you have had to move the custom reqId string generation into a dedicated binding prop

Actually, the reason is I messed up. I must have been using `fastify.log` rather than `req.log` when testing :man_facepalming: 

It hardly matters which we use for this endpoint, since there's just one log. If we use lots of logs (like in Shaun's PR), it's cleaner to create a child logger.

raisedadead commented:
Yeah - I see the confusion. I'd say lets stick to one pattern across the codebase.

My preference would be:

- For routes, define the child logger early on in file and use it. Exactly like Shaun's PR or use `fastify.log.child({ req }).debug('message specific to a request')` as a short hand as needed.
- For other areas, use `fastify.log.debug('message in some chunk of code that is generic like db connection, sentry boot etc')` etc.

and so on for `warn` and `info`, `error` etc.

ojeytonwilliams committed:
refactor: use child logger

raisedadead commented on api/src/server.ts:
```diff
@@ -13,16 +17,29 @@ import {
+  const hostname = req.hostname || 'HOSTNAME not found';
+  const remotePort = req.socket.remotePort || 'REMOTE_PORT not found';
+
+  return {
```
```suggestion
  return {
    REQ_ID: req.id,
```

Include the request ID

raisedadead commented on api/src/server.ts:
```diff
@@ -38,12 +55,14 @@ const envToLogger = {
     serializers: {
       req: requestSerializer
```
```suggestion
      req: (req: FastifyRequest) => {
        return {
          method: req.method,
          url: req.url
        };
      }
```

The `requestSerializer` is way too noisy for local development. Parity is not a concern here.

raisedadead commented on api/src/routes/public/status.ts:
```diff
@@ -15,25 +14,7 @@ export const statusRoute: FastifyPluginCallbackTypebox = (
-        QUERY: query
-      })
-      .debug('returning a ping');
+    req.log.child({ req }).debug('returning a ping');
```
```suggestion
    fastify.log.child({ req }).debug('returning a ping');
```

This is not better, but is a clearer mental model. Use `fastify.log` and `fastify.log.child` everywhere.

raisedadead reviewed (changes requested):
Adjustments.

ojeytonwilliams commented:
Hey @raisedadead, in an ideal world we should figure out how to filter the logs, turn off `disableRequestLogging`, use `req.log.debug('message')` in routes and `fastify.log.debug('message')` outside them. I think that's a simple mental model and keeps the code clean. 

That being said, this has felt like bikeshedding for a while now, and we need to do something, so I'm fine with adopting your latest suggestion.

ojeytonwilliams committed:
refactor: use fastify.log + simplify dev logging

Co-authored-by: Mrugesh Mohapatra <1884376+raisedadead@users.noreply.github.com>

raisedadead commented:
> use req.log.debug('message') in routes and fastify.log.debug('message') outside them 

&

> I think that's a simple mental model and keeps the code clean.

IMHO, these two are contradictory statements to me. 

I keep preaching the tools need to work for us not the other way around. I need the logging to work exactly like I need it -- that is I do not need to look at all the requests (so unlikely we will ever use full request logging)

I feel like puking when I see `req.log`, It looks weird and doesn't come first to my mind when I am debugging in the real world. I can see how `req.log` might feel natural when someone is dedicatedly working only on the API and not the larger ecosystem (parsing logs in docker, journalctl, etc.) and I get that its a quirk of fastify. 

I just refuse to accept that. :) Sue me.

I am not very pleased with `fastify.log` either, but it will do. Someday, I may tear it down and use standard `logfmt` based logging utils which will actually help out o11y goals.

raisedadead reviewed (approved):
LGTM - Thanks for your patience with my review.

ojeytonwilliams commented:
No worries, it needed discussing. Let's see how this works in practice and tweak based on that.

ShaunSHamilton reviewed (approved):
LOTM 👍
